fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,234 +4,268 @@
-from typing import Any, Dict, Generator, List, Type, cast
+import contextlib
+from itertools import chain
+import logging
+from typing import Any, Callable, Generator, List, Tuple, Union
 
 import torch
-from torch import Tensor, nn
+from torch import nn
+from torch.autograd import Variable
 import torch.distributed as dist
 from torch.nn import Parameter
 
 from fairscale.optim import OSS
+from fairscale.optim.utils import Workhandle
 
 
 class ShardedDataParallel(nn.Module):
-    """Implements distributed data parallel training with optimizer state sharding.
-
-    A simplified version of :class:`torch.nn.parallel.DistributedDataParallel`.
-    This version uses a c10d process group for communication and optionally
-    broadcast buffers.
-
-    Args:
-        module (~torch.nn.Module): module to be parallelized
-        optimizer (~torch.optim.Optimizer): optimizer to be used for training
-        optimizer_params(Dict): extra parameters for the optimizer
-        world_size (int): number of parallel workers
-        broadcast_buffers (bool): flag that enables syncing (broadcasting) buffers of
-        the module at beginning of the forward function. (default: ``True``)
-        process_group (optional): the c10d process group to be used for
-            distributed gradient reduction. If None, the default WORLD process group
-            will be used.
-        buffer_size (int, optional): number of elements to buffer before
-            performing reduce (default: 512k). Used to reduce multiple small
-            params to avoid communication overhead.
+    """
+    Wrap the model, and reduce the gradients to the right rank during the backward pass.
+
+    - the partition is given by the sharded optimizer
+    - wrap the base model with a model which knows where to reduce each gradient
+    - add an autograd function which calls the model grad dispatch on the way back
+
+     Args:
+        module (nn.Module):
+            model to be wrapped
+        sharded_optimizer (OSS, or list of OSS):
+            the sharded optimizer(s) which will decide the gradient partitioning
+
+    Keyword Args:
+        process_group (torch.nn.Optimizer):
+            optimizer to shard (default: SGD)
+        process_group (group):
+            torch.distributed group (default: group.WORLD)
+        broadcast_buffers (bool):
+            whether to broadcast model buffers in between ranks at the beginning of each forward pass
+
     """
 
     def __init__(
         self,
         module: nn.Module,
-        optimizer: Type[torch.optim.Optimizer],
-        optimizer_params: Dict[str, Any],
-        world_size: int,
-        broadcast_buffers: bool,
+        sharded_optimizer: Union[OSS, List[OSS]],
         process_group: Any = None,
-        buffer_size: int = 2 ** 19,
+        broadcast_buffers: bool = True,
     ):
         super().__init__()
 
-        self.module = module
-        self.world_size = world_size
+        self.base_model = module
```

Comment:
Would it be possible to change `self.base_model` to `self.module` as well? This would keep it consistent (this is also how we access base class across DDP etc in lightning).

Reply:
done